### PR TITLE
[SPARK-40152][SQL][TESTS] Add tests for SplitPart

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Add tests for `SplitPart`.

### Why are the changes needed?

Improve test coverage.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

N/A.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala [hunk 1]
```diff
@@ -2522,4 +2522,24 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
           Date.valueOf("2017-02-12")))
     }
   }
+
+  test("SplitPart") {
```
Comment (wangyum) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2526]: It seems can't directly test `SplitPart`. So just test `replacement`:
https://github.com/apache/spark/blob/b07ef74feb44d6f33538a47bdb54fc021b4a40d3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L2864-L2866


```scala
checkEvaluation(SplitPart(Literal.create("11.12.13", StringType), Literal.create(".", StringType), Literal(3)), "13")
```

```
Caused by: org.apache.spark.SparkUnsupportedOperationException: [INTERNAL_ERROR] Cannot evaluate expression: split_part(11.12.13, ., 3)
	at org.apache.spark.sql.errors.QueryExecutionErrors$.cannotEvaluateExpressionError(QueryExecutionErrors.scala:71)
	at org.apache.spark.sql.catalyst.expressions.RuntimeReplaceable.eval(Expression.scala:369)
	at org.apache.spark.sql.catalyst.expressions.RuntimeReplaceable.eval$(Expression.scala:368)
	at org.apache.spark.sql.catalyst.expressions.SplitPart.eval(stringExpressions.scala:2859)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -2270,7 +2270,7 @@ case class ElementAt(
               case Some(value) =>
                 val defaultValueEval = value.genCode(ctx)
                 s"""
-                  ${defaultValueEval.code};
+                  ${defaultValueEval.code}
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:2273]: is this a bug fix?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala [hunk 3]
```diff
@@ -2522,4 +2522,24 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
...
+    checkEvaluation(ElementAt(str, Literal(1), outOfBoundValue), UTF8String.fromString("11"))
+    checkEvaluation(ElementAt(str, Literal(10), outOfBoundValue), UTF8String.fromString(""))
+    checkEvaluation(ElementAt(str, Literal(-10), outOfBoundValue), UTF8String.fromString(""))
+
+    checkEvaluation(ElementAt(StringSplitSQL(Literal.create(null, StringType), delimiter),
+      Literal(1), outOfBoundValue), null)
+    checkEvaluation(ElementAt(StringSplitSQL(Literal.create("11.12.13", StringType),
+      Literal.create(null, StringType)), Literal(1), outOfBoundValue), null)
+
+    intercept[Exception] {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2541]: Can we catch narrower exception?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 2 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:2273]: No. Please see: https://github.com/apache/spark/pull/37589#pullrequestreview-1081493528

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2541]: Do you mean `TestFailedException`?

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2541]: org.scalatest.exceptions.TestFailedException is scala test exception, not Spark exception. It seems meaningless:
```scala
    intercept[SparkRuntimeException] {
      checkEvaluation(ElementAt(str, Literal(0), outOfBoundValue), null)
    }.getMessage.contains("The index 0 is invalid")
```

```
Expected exception org.apache.spark.SparkRuntimeException to be thrown, but org.scalatest.exceptions.TestFailedException was thrown
ScalaTestFailureLocation: org.apache.spark.sql.catalyst.expressions.CollectionExpressionsSuite at (CollectionExpressionsSuite.scala:2543)
org.scalatest.exceptions.TestFailedException: Expected exception org.apache.spark.SparkRuntimeException to be thrown, but org.scalatest.exceptions.TestFailedException was thrown
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
	at org.scalatest.funsuite.AnyFunSuite.newAssertionFailedException(AnyFunSuite.scala:1563)
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2526]: Can we simply add more tests in the existing test case `test("elementAt")`? We just need to test the `outOfBoundValue`

**Comment by srowen** [comment]
Merged to master/3.3

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala [hunk 4]
```diff
@@ -2522,4 +2522,24 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
...
+      checkEvaluation(ElementAt(str, Literal(0), outOfBoundValue), null)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2542]: We have a dedicated function `checkExceptionInExpression` for testing the expression error behavior.

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2542]: Actually this is already tested in `test("elementAt")`

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2542]: See https://github.com/apache/spark/pull/37626/files#r952639395

I think we should just expand the existing `test("elementAt")`.

**Comment by srowen** [comment]
Ah sorry did it again, merged this already. If it's important make another PR and I'll merge it

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala (hunk 4 above)
Comment (wangyum) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2542]: https://github.com/apache/spark/pull/37637

**Pull request closed without merging**